yb-voyager/src/tgtdb/conn_pool.go

```diff
@@ -44,27 +45,94 @@ type ConnectionPool struct {
-		conns:                     make(chan *pgx.Conn, params.NumConnections),
-		connIdToPreparedStmtCache: make(map[uint32]map[string]bool, params.NumConnections),
+		conns:                     make(chan *pgx.Conn, params.NumMaxConnections),
+		idleConns:                 make(chan *pgx.Conn, params.NumMaxConnections),
+		connIdToPreparedStmtCache: make(map[uint32]map[string]bool, params.NumMaxConnections),
 		disableThrottling:         false,
+		size:                      params.NumConnections,
+		pendingConnsToClose:       0,
+	}
+	for i := 0; i < params.NumMaxConnections; i++ {
+		pool.idleConns <- nil
 	}
 	for i := 0; i < params.NumConnections; i++ {
-		pool.conns <- nil
+		c := <-pool.idleConns
+		pool.conns <- c
 	}
 	if pool.params.SessionInitScript == nil {
 		pool.params.SessionInitScript = defaultSessionVars
 	}
 	return pool
 }
 
+func (pool *ConnectionPool) GetNumConnections() int {
+	return pool.size
+}
+
+func (pool *ConnectionPool) UpdateNumConnections(delta int) error {
+	pool.pendingConnsToCloseLock.Lock()
+	defer pool.pendingConnsToCloseLock.Unlock()
+
+	newSize := pool.size + delta
+
+	if newSize < 1 || newSize > pool.params.NumMaxConnections {
+		return fmt.Errorf("invalid new pool size %d. "+
+			"Must be between 1 and %d", newSize, pool.params.NumMaxConnections)
+	}
+
+	if delta == 0 {
+		log.Infof("adaptive: No change in pool size. Current size is %d", pool.size)
+		return nil
+	}
+	if delta > 0 {
+		// for increases, process them synchronously.
+		// If there is non-zero pendingConnsToClose, then we reduce that count.
+		if pool.pendingConnsToClose >= 0 {
+			// since we are increasing, we can just reduce the pending close count
+			if pool.pendingConnsToClose > delta {
+				log.Infof("adaptive: Decreasing pendingConnsToClose by %d", delta)
+				pool.pendingConnsToClose -= delta
+				delta = 0
+			} else {
+				log.Infof("adaptive: Decreasing pendingConnsToClose to 0 by %d", pool.pendingConnsToClose)
+				delta -= pool.pendingConnsToClose
+				pool.pendingConnsToClose = 0
+			}
+		}
+		// Additionally, pick conns from the idle pool, and add it to the main pool.
+		for i := 0; i < delta; i++ {
+			conn := <-pool.idleConns
+			pool.conns <- conn
+		}
+		log.Infof("adaptive: Added %d new connections. Pool size is now %d", delta, newSize)
+	} else {
+		// for decreases, process them asynchronously.
+		// The problem with processing them synchronously is that we may have to wait for the
+		// connections to be returned to the pool. Not only that, this goroutine that is trying to
+		// update the pool size by reading from the channel is also competing with multiple
+		// other goroutines that are trying to ingest data.
+		//  This might take significant time depending on the query execution time and the lock contention
+		//  So, instead, we just register the request to close the connections,
+		// and the connections are returned to the idle pool when the query execution is done.
+		pool.pendingConnsToClose += -delta
```

Comment:
Use `Math.Abs()`
https://pkg.go.dev/math#Abs

Reply:
Unfortunately go doesn't seem to have an abs function for int. This is for float64 only.